CesiumGltf/include/CesiumGltf/Model.h

```diff
@@ -19,6 +22,61 @@ struct CESIUMGLTF_API Model : public ModelSpec {
    */
   void merge(Model&& rhs);
 
+  /**
+   * @brief A callback function for {@link forEachPrimitiveInScene}.
+   */
+  typedef void ForEachPrimitiveInSceneCallback(
+      Model& gltf,
+      Node& node,
+      Mesh& mesh,
+      MeshPrimitive& primitive,
+      const glm::dmat4& transform);
+
+  /**
+   * @brief Apply the given callback to all relevant primitives.
+   *
+   * If the given `sceneID` is non-negative and exists in the given glTF,
+   * then the given callback will be applied to all meshes of this scene.
+   *
+   * If the given `sceneId` is negative, then the meshes that the callback
+   * will be applied to depends on the structure of the glTF model:
+   *
+   * * If the glTF model has a default scene, then it will
+   *   be applied to all meshes of the default scene.
+   * * Otherwise, it will be applied to all meshes of the the first scene.
+   * * Otherwise (if the glTF model does not contain any scenes), it will
+   *   be applied to all meshes that can be found by starting a traversal
+   *   at the root node.
+   * * Otherwise (if there are no scenes and no nodes), then all meshes
+   *   will be traversed.
+   *
+   * @param sceneID The scene ID (index)
+   * @param callback The callback to apply
+   */
+  void forEachPrimitiveInScene(
+      int sceneID,
+      std::function<ForEachPrimitiveInSceneCallback>&& callback);
+
+  /**
+   * @brief A callback function for {@link forEachPrimitiveInScene}.
+   */
+  typedef void ForEachPrimitiveInSceneConstCallback(
+      const Model& gltf,
+      const Node& node,
+      const Mesh& mesh,
+      const MeshPrimitive& primitive,
+      const glm::dmat4& transform);
+
+  /** @copydoc Gltf::forEachPrimitiveInScene() */
+  void forEachPrimitiveInScene(
+      int sceneID,
+      std::function<ForEachPrimitiveInSceneConstCallback>&& callback) const;
+
+  /**
+   * @brief Fills in smooth normals for any primitives with missing normals.
+   */
+  void generateMissingNormalsSmooth();
```

Comment:
I just notice that this generates normal for every gltf primitives in the model. It maybe problematic here since some primitives may requires tangent and bitangent too, which will require flat normals in those cases. For example, imagine a scene has two primitives with both missing normals and tangents too. However, one primitive has material with no-normal mapping. In this case, it doesn't requires tangent and bitangent and therefore, generating smooth normal should be okay. However, another primitive has normal mapping in the material. It will require tangent to be generated and flat normal needs to be used here, instead of a smooth normal

Reply:
> some primitives may requires tangent and bitangent too, which will require flat normals in those cases.

I could be missing something obvious, but why do tangents and bitangents necessitate flat normals?